core/src/sumeragi/fault.rs

```diff
@@ -39,865 +37,1333 @@ impl FaultInjection for NoFault {
+                state_machine_guard
+                    .wsv
+                    .peers()
+                    .iter()
+                    .map(|peer| peer.id().clone())
+                    .collect(),
+            )
+            .build(0)
+            .expect("Should be able to reconstruct topology from wsv");
+    } else {
+        // We need to perform a round of some form
+
+        let maybe_genesis_network = {
+            let mut state_machine_guard = sumeragi
+                .sumeragi_state_machine_data
+                .lock()
+                .expect("take lock");
+            state_machine_guard.genesis_network.take()
+        };
 
-        match msg {
-            SumeragiMessage(data) => self.broker.issue_send(data.into_v1()).await,
-            BlockSync(data) => self.broker.issue_send(data.into_v1()).await,
-            Health => {}
+        if let Some(genesis_network) = maybe_genesis_network {
+            sumeragi_init_commit_genesis(sumeragi, genesis_network);
+        } else {
+            sumeragi_init_listen_for_genesis(sumeragi, &mut incoming_message_receiver);
         }
     }
-}
 
-impl<G: GenesisNetworkTrait, F: FaultInjection> SumeragiWithFault<G, F> {
-    /// Initializes sumeragi with the `latest_block_hash` and `block_height` after Kura loads the blocks.
-    pub fn init(&mut self, latest_block: HashOf<VersionedCommittedBlock>, block_height: u64) {
-        self.block_height = block_height;
-        self.topology.apply_block(latest_block);
+    {
+        let state_machine_guard = sumeragi.sumeragi_state_machine_data.lock().unwrap();
+        assert!(state_machine_guard.latest_block_height >= 1);
+        assert_eq!(
+            state_machine_guard.latest_block_hash,
+            state_machine_guard.wsv.latest_block_hash()
+        );
+        trace!(
+            "I, {}, finished sumeragi init. My role in the next round is {:?}",
+            sumeragi.peer_id.public_key,
+            state_machine_guard.current_topology.role(&sumeragi.peer_id),
+        );
     }
 
-    /// Updates network topology by taking the actual list of peers from `WorldStateView`.
-    /// Updates it only if there is a change in WSV peers, otherwise leaves the order unchanged.
-    #[allow(clippy::expect_used, clippy::unused_async)]
-    pub async fn update_network_topology(&mut self) {
-        let wsv_peers: HashSet<_> = self.wsv.trusted_peers_ids().clone().into_iter().collect();
-        let topology_peers: HashSet<_> = self.topology.sorted_peers().iter().cloned().collect();
-        if topology_peers != wsv_peers {
-            self.topology = self.topology
-                .clone()
-                .into_builder()
-                .with_peers(wsv_peers)
-                .build()
-                // TODO: Check it during instruction execution.
-                .expect("The safety of changing the number of peers should have been checked at Instruction execution stage.");
-        }
-    }
+    // do normal rounds
```

Comment:
Aren't the purpose of comments to inform you about the code? Plus if it's a function you loose information because you don't know where it is called from. Here it is obvious.